src/pages/Compare/index.tsx

```diff
@@ -62,43 +70,66 @@ export class CompareBase extends React.Component<Props> {
   }
 
   loadData(prevProps?: Props) {
-    const { match } = this.props;
+    const { match, path } = this.props;
     const { addonId, baseVersionId, headVersionId } = match.params;
 
     if (
       !prevProps ||
+      (prevProps.path && path !== prevProps.path) ||
```

Comment:
I guess this is fine but relying on previous props still feels a little fragile to me.

I wanted to present an example of relying on Redux state instead of previous props but you don't necessarily need to make this change. I want to show this because I don't think it's any more or less complex than what you already have. In other words, relying on previous props may not be simpler than just relying on Redux state.

This patch mostly works but it doesn't handle error states since there's no loading flag in the state. It will probably infinitely fetch data on error but that could be fixed by storing a loading flag in the state.

```diff
diff --git a/src/pages/Compare/index.tsx b/src/pages/Compare/index.tsx
index 0e33d7a..c3e8944 100644
--- a/src/pages/Compare/index.tsx
+++ b/src/pages/Compare/index.tsx
@@ -70,23 +70,27 @@ export class CompareBase extends React.Component<Props> {
   }
 
   loadData(prevProps?: Props) {
-    const { match, path } = this.props;
-    const { addonId, baseVersionId, headVersionId } = match.params;
+    const { compareInfo, match, path } = this.props;
+
+    const addonId = parseInt(match.params.addonId, 10);
+    const baseVersionId = parseInt(match.params.baseVersionId, 10);
+    const headVersionId = parseInt(match.params.headVersionId, 10);
 
     if (
-      !prevProps ||
-      (prevProps.path && path !== prevProps.path) ||
-      addonId !== prevProps.match.params.addonId ||
-      baseVersionId !== prevProps.match.params.baseVersionId ||
-      headVersionId !== prevProps.match.params.headVersionId
+      compareInfo === null ||
+      (compareInfo &&
+        (compareInfo.path !== path ||
+          compareInfo.addonId !== addonId ||
+          compareInfo.baseVersionId !== baseVersionId ||
+          compareInfo.headVersionId !== headVersionId))
     ) {
       const { dispatch, _fetchDiff } = this.props;
 
       dispatch(
         _fetchDiff({
-          addonId: parseInt(addonId, 10),
-          baseVersionId: parseInt(baseVersionId, 10),
-          headVersionId: parseInt(headVersionId, 10),
+          addonId,
+          baseVersionId,
+          headVersionId,
           path: path || undefined,
         }),
       );
diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index fcf8e64..a776909 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -190,7 +190,11 @@ export type VersionsMap = {
 };
 
 export type CompareInfo = {
+  addonId: number;
   diffs: DiffInfo[];
+  path: string;
+  baseVersionId: number;
+  headVersionId: number;
   mimeType: string;
 };
 
@@ -220,7 +224,7 @@ export const actions = {
       baseVersionId: number;
       headVersionId: number;
       version: ExternalVersionWithDiff;
-      path?: string;
+      path: string;
     }) => resolve(payload);
   }),
 };
@@ -231,7 +235,7 @@ export type VersionsState = {
   };
   compareInfo:
     | CompareInfo // data successfully loaded
-    | null // an error has occured
+    | null // data has not yet loaded or an error occurred
     | undefined; // data not fetched yet
   versionInfo: {
     [versionId: number]: Version;
@@ -245,7 +249,7 @@ export type VersionsState = {
 
 export const initialState: VersionsState = {
   byAddonId: {},
-  compareInfo: undefined,
+  compareInfo: null,
   versionFiles: {},
   versionInfo: {},
 };
@@ -535,7 +539,7 @@ export const fetchDiff = ({
 
     dispatch(actions.beginFetchDiff());
 
-    const response = await _getDiff({
+    const externalVersion = await _getDiff({
       addonId,
       apiState,
       baseVersionId,
@@ -543,18 +547,18 @@ export const fetchDiff = ({
       path,
     });
 
-    if (isErrorResponse(response)) {
-      _log.error(`TODO: handle this error response: ${response.error}`);
+    if (isErrorResponse(externalVersion)) {
+      _log.error(`TODO: handle this error response: ${externalVersion.error}`);
       dispatch(actions.abortFetchDiff());
     } else {
-      dispatch(actions.loadVersionInfo({ version: response }));
+      dispatch(actions.loadVersionInfo({ version: externalVersion }));
       dispatch(
         actions.loadDiff({
           addonId,
           baseVersionId,
           headVersionId,
-          path,
-          version: response,
+          path: path || externalVersion.file.selected_file,
+          version: externalVersion,
         }),
       );
     }
@@ -630,7 +634,13 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
       };
     }
     case getType(actions.loadDiff): {
-      const { baseVersionId, headVersionId, version } = action.payload;
+      const {
+        addonId,
+        baseVersionId,
+        headVersionId,
+        path,
+        version,
+      } = action.payload;
 
       const { entries, selectedPath } = getVersionInfo(state, headVersionId);
       const entry = entries.find((e) => e.path === selectedPath);
@@ -643,12 +653,16 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
       return {
         ...state,
         compareInfo: {
+          addonId,
+          baseVersionId,
+          headVersionId,
           diffs: createInternalDiffs({
             baseVersionId,
             headVersionId,
             version,
           }),
           mimeType: entry.mimeType,
+          path,
         },
       };
     }
```

Reply:
> I guess this is fine but relying on previous props still feels a little fragile to me.

I do agree. I believe this could be fixed by #363, which I am willing to take later this quarter.

Reply:
`useEffect` will help cut down on the tests but you'll still need to make comparisons to something to decide when to fetch data. In `useEffect`, the approach of comparing to previous props may not be possible. 

But, yeah, we can revisit this later. I'm slightly worried that https://github.com/mozilla/addons-code-manager/issues/286 won't work with this implementation but we'll find out soon enough!